Move the CompatibilityMode type to the lib package

na-- opened the pull request:
This was needed because if the `lib` package depended on the `js/compiler` package, any Go project imports lib would also import goja and babel...

@imiric, @MStoykov, do you remember why a `null.String` was used in the `RuntimeOptions`, instead of the `CompatibilityMode` type directly?

na-- commented:
Ah... well then, yeah, I can refactor this fully so we just use the simple type... I probably should have insisted on it in the original PR, and would have, had I realized that it would pull `goja` in `lib/` :sweat_smile: 

I don't think refactoring this now it's going to be an issue, considering that `k6-0.26.0 archive github.com/loadimpact/k6/samples/http_get.js  && (tar xfO archive.tar metadata.json | jq '.compatibilityMode')` would be `"extended"`. So we're not using `""` as the default version anyway, it's either going to be `null` (i.e. not present, in old archives) or `"extended"` or `"base"`

MStoykov commented:
I [wanted](https://github.com/loadimpact/k6/pull/1206#discussion_r336356477) the 0 value to be invalid so this might be a problem, but you can just change it to 0 being extended ;)

na-- commented:
> I'm up for changing it if it doesn't cause any issues.

Unfortunately, it does... I changed it, but then when I refactored the tests I found a minor bug (present both in the current k6 and in my changes) that couldn't be fixed without a `null`-able type... The bug is that k6 should be able to overwrite the compatibility mode it gets from an archive, i.e. this should return an error:
```
k6 run --compatibility-mode=base es6_archive_created_with_extended_compat_mode.tar
```

So I'll leave the `null.String` for now, just fix the minor issues I found and add a few more config tests.

na-- commented:
This is done from my perspective, please take another look. You can read the body of the latest commit to get a better idea about all of the changes that happen in it.

MStoykov commented on cmd/runtime_options_test.go:
```diff
@@ -24,246 +24,255 @@ import (
+		))
+	}
+	jsCodeLines = append(jsCodeLines, "}")
+	jsCode := []byte(strings.Join(jsCodeLines, "\n"))
```
Maybe use [bytes.Buffer](https://golang.org/pkg/bytes/#Buffer) or [string.Builder](https://golang.org/pkg/strings/#Builder) ? I think [bytes.Buffer](https://golang.org/pkg/bytes/#Buffer) with [fmt.Fprintf](https://yourbasic.org/golang/build-append-concatenate-strings-efficiently/#before-go-1-10) will be the best from performance(not that it matters all that much in this case and readability :).

na-- replied:
Fixed by https://github.com/loadimpact/k6/pull/1325/commits/c9d75bd1cb74cafd11182125da5ffe2d8e61b37b

MStoykov commented on cmd/runtime_options_test.go:
```diff
@@ -24,246 +24,255 @@ import (
-			require.NoError(t, err)
+	archive := runner.MakeArchive()
+	archiveBuf := &bytes.Buffer{}
+	assert.NoError(t, archive.Write(archiveBuf))
```
I think this should be `require` as if this fails the below ones will definitely fail and we won't get any useful information from their failure

na-- replied:
Fixed by https://github.com/loadimpact/k6/pull/1325/commits/c9d75bd1cb74cafd11182125da5ffe2d8e61b37b

imiric commented on cmd/runtime_options.go:
```diff
@@ -63,24 +67,42 @@ extended: base + Babel with ES2015 preset + core.js v2,
+			opts.IncludeSystemEnvVars = null.BoolFrom(val)
+		}
+	}
+
```
I realize this fixes the handling of these env vars, but the special treatment of these specific ones feels like we're adding more things to iron out in #883. Hopefully we can have a saner way of dealing with these fallback scenarios.

No action needed, just a comment. :)

na-- replied:
I have a vague idea how to implement that logic in a generic way, when we get to that... It's not as if the fallback/overriding logic is very complicated, I don't even consider this "special treatment" - it is all pretty straightforward "if this is not set at the top tier, try the bottom one"... 

So, I think we would be able to implement something to handle all of the different types and cases, that is relatively simple, generic, and easy to test, even with the limitations imposed by Go.

na-- commented:
@MStoykov @imiric any idea about the AppVeyor error? This code was in the original test: https://github.com/loadimpact/k6/blob/a1f38026cbf0c66d15b8e60c904f4dd4d031441a/cmd/runtime_options_test.go#L223-L229

but now that I don't use `os.ClearEnv()`, it should have been unnecessary :confused:

na-- commented:
Turns out there was another `os.ClearEnv()` call in a test of the same package, without any restoring of the cleared environment variables after it :man_facepalming: :crossed_fingers: that fixing it (and another case in `lib/`) would be sufficient to not trigger the strange windows error...
